library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
+
+    var shift =
+      inlineNumberOfLeadingZeros(blo, bhi) - inlineNumberOfLeadingZeros(alo, ahi)
+    val initialBShift = inline_<<(blo, bhi, shift)
+    var bShiftLo = initialBShift._1
+    var bShiftHi = initialBShift._2
+    var remLo = alo
+    var remHi = ahi
+    var quotLo = 0
+    var quotHi = 0
+
+    /* Invariants:
+     *   bShift == b << shift == b * 2^shift
+     *   quot >= 0
+     *   0 <= rem < 2 * bShift
+     *   quot * b + rem == a
+     */
+    while (shift >= 0 && (remHi & 0xffe00000) != 0) {
+      if (inlineUnsigned_>=(remLo, remHi, bShiftLo, bShiftHi)) {
+        val newRem = inline_-(remLo, remHi, bShiftLo, bShiftHi)
+        remLo = newRem._1
+        remHi = newRem._2
+        if (shift < 32)
+          quotLo |= (1 << shift)
+        else
+          quotHi |= (1 << shift) // == (1 << (shift - 32))
+      }
+      shift -= 1
+      val newBShift = inline_>>>(bShiftLo, bShiftHi, 1)
+      bShiftLo = newBShift._1
+      bShiftHi = newBShift._2
+    }
 
-    masked(c0, c1n, c2 + (c1n >> BITS))
-  }
+    // Now rem < 2^53, we can finish with a double division
+    if (inlineUnsigned_>=(remLo, remHi, blo, bhi)) {
+      val remDouble = remHi * TwoPow32 + remLo.toUint
+      val bDouble = bhi * TwoPow32 + blo.toUint
+
+      if (ask != AskRemainder) {
+        val rem_div_bDouble = remDouble / bDouble
+        val newQuot = inline_+(quotLo, quotHi,
+            rawToInt(rem_div_bDouble), rawToInt(rem_div_bDouble / TwoPow32))
+        quotLo = newQuot._1
+        quotHi = newQuot._2
+      }
 
-  def /(y: RuntimeLong): RuntimeLong = (x divMod y)(0)
-  def %(y: RuntimeLong): RuntimeLong = (x divMod y)(1)
+      if (ask != AskQuotient) {
+        val rem_mod_bDouble = remDouble % bDouble
+        remLo = rawToInt(rem_mod_bDouble)
+        remHi = rawToInt(rem_mod_bDouble / TwoPow32)
+      }
+    }
 
-  //override def getClass(): Class[Long] = null
+    if (ask == AskQuotient) new RuntimeLong(quotLo, quotHi)
+    else if (ask == AskRemainder) new RuntimeLong(remLo, remHi)
+    else js.Tuple4(quotLo, quotHi, remLo, remHi)
+  }
+
+  // Support for intrinsics
 
   def toBinaryString: String = {
-    val zeros = "0000000000000000000000" // 22 zeros
-    @inline def padBinary22(i: Int) = {
+    val zeros = "00000000000000000000000000000000" // 32 zeros
+    @inline def padBinary32(i: Int) = {
       val s = Integer.toBinaryString(i)
       zeros.substring(s.length) + s
     }
 
-    if (h != 0) Integer.toBinaryString(h) + padBinary22(m) + padBinary22(l)
-    else if (m != 0) Integer.toBinaryString(m) + padBinary22(l)
-    else Integer.toBinaryString(l)
+    val lo = this.lo
+    val hi = this.hi
+
+    if (hi != 0) Integer.toBinaryString(hi) + padBinary32(lo)
+    else Integer.toBinaryString(lo)
   }
 
   def toHexString: String = {
-    val zeros = "000000" // 6 zeros
-    @inline def padHex(i: Int, len: Int) = {
+    val zeros = "00000000" // 8 zeros
+    @inline def padBinary8(i: Int) = {
```

Comment:
This should be called `padHex8`. But feel free to inline these. They don't make a lot of sense anymore since they are used only once.